### ISSUE TITLE
`offset` / `limit` returning invalid results with `where` filtering

### ISSUE THREAD

**Issue by Tenrys** [neo4j/graphql#2396]
**Describe the bug**
Offset / limit don't seem to work properly, when used in accordance with filters, returning 0 results past a certain point when I know I should be getting some.

**Type definitions**
Isolated type definitions that cause the issue for me:

```graphql
type PostalCode @exclude(operations: [DELETE]) {
  archivedAt: DateTime
  number: String! @id(autogenerate: false)
  
  address: [Address!]! @relationship(type: "HAS_POSTAL_CODE", direction: IN)
}
extend type PostalCode
  @auth(
    rules: [
      { where: { archivedAt: null } }
    ]
  )

union AddressNode = Estate | Customer | Demand | Deed

type Address @exclude(operations: [DELETE]) {  
  archivedAt: DateTime
  uuid: ID! @id
  createdAt: DateTime! @timestamp(operations: [CREATE])
  updatedAt: DateTime! @timestamp(operations: [CREATE, UPDATE])

  address: String! @unique
  streetNumber: String
  route: String! @coalesce(value: "")
  postalCode: PostalCode! @relationship(type: "HAS_POSTAL_CODE", direction: OUT)
  locality: String! @coalesce(value: "")
  administrativeAreaLevel1: String! @coalesce(value: "")
  administrativeAreaLevel2: String
  country: String! @coalesce(value: "")
  location: Point!
  
  node: [AddressNode!]! @relationship(type: "HAS_ADDRESS", direction: IN)
}
extend type Address
  @auth(
    rules: [
      { where: { archivedAt: null } }
    ]
  )

type Mandate @exclude(operations: [DELETE]) {
  archivedAt: DateTime
  number: String! @unique # numéro
  createdAt: DateTime! @timestamp(operations: [CREATE])
  updatedAt: DateTime! @timestamp(operations: [CREATE, UPDATE])

  price: Float!
  
  valuation: Valuation! @relationship(type: "HAS_VALUATION", direction: OUT)
}
extend type Mandate
  @auth(
    rules: [
      { where: { archivedAt: null } }
    ]
  )

type Valuation @exclude(operations: [DELETE]) {
  archivedAt: DateTime
  uuid: ID! @id
  createdAt: DateTime! @timestamp(operations: [CREATE])
  updatedAt: DateTime! @timestamp(operations: [CREATE, UPDATE])

  estate: Estate @relationship(type: "VALUATION_FOR", direction: OUT)
}
extend type Valuation
  @auth(
    rules: [
      { where: { archivedAt: null } }
    ]
  )
  
enum EstateType {
  APARTMENT
  HOUSE_VILLA
  TOWNHOUSE
  LAND
  PARKING
  BOX
  BUILDING
  COMMERCIAL_PREMISE
  CHALET
  CASTLE
  OFFICE
  BUSINESS_FUND
}
type Estate @exclude(operations: [DELETE]) {
  archivedAt: DateTime
  uuid: ID! @id
  createdAt: DateTime! @timestamp(operations: [CREATE])
  updatedAt: DateTime! @timestamp(operations: [CREATE, UPDATE])

  estateType: EstateType!  
  area: Float!
  floor: Int
  
  address: Address @relationship(type: "HAS_ADDRESS", direction: OUT)
}
extend type Estate
  @auth(
    rules: [
      { where: { archivedAt: null } }
    ]
  )
```

**To Reproduce**

1. Import [this JSON data](https://github.com/neo4j/graphql/files/9971436/response.txt), matching the types described above
2. Run this Query:

```graphql
query Mandates($where: MandateWhere, $options: MandateOptions) {
  mandates(options: $options, where: $where) {
    valuation { 
      estate { 
        uuid 
      } 
    }
  }
}
```

with these variables:

```json
{
  "options": {},
  "where": {
    "price_GTE": 0,
    "valuation": {
      "estate": {
        "address": {
          "postalCode": {
            "number_IN": [
              "13001"
            ]
          }
        },
        "area_GTE": 0,
        "estateType_IN": [
          "APARTMENT"
        ],
        "floor_GTE": 0
      }
    }
  }
}
```

You should get about 26 results there.

If you try adding `"offset": 0` and `"limit": 20` to the options, you should get 20 results, accordingly.

However, if you set `offset` to 20 and `limit` to 40, you won't get any returned results.

From `DEBUG=@neo4j/graphql:*`:

```
Cypher:
MATCH (this:`Mandate`)
WHERE ((this.price >= $param0 AND EXISTS {
    MATCH (this)-[:HAS_VALUATION]->(this0:`Valuation`)
    WHERE EXISTS {
        MATCH (this0)-[:VALUATION_FOR]->(this1:`Estate`)
        WHERE (this1.area >= $param1 AND this1.floor >= $param2 AND this1.estateType IN $param3 AND EXISTS {
            MATCH (this1)-[:HAS_ADDRESS]->(this2:`Address`)
            WHERE EXISTS {
                MATCH (this2)-[:HAS_POSTAL_CODE]->(this3:`PostalCode`)
                WHERE this3.number IN $param4
            }
        })
    }
}) AND this.archivedAt IS NULL)

WITH *

SKIP $param5
LIMIT $param6
CALL {
    WITH this
    MATCH (this)-[this4:HAS_VALUATION]->(this_valuation:`Valuation`)
    WHERE this_valuation.archivedAt IS NULL
    CALL {
        WITH this_valuation
        MATCH (this_valuation)-[this5:VALUATION_FOR]->(this_valuation_estate:`Estate`)
        WHERE this_valuation_estate.archivedAt IS NULL
        WITH this_valuation_estate { .uuid } AS this_valuation_estate
        RETURN head(collect(this_valuation_estate)) AS this_valuation_estate
    }
    WITH this_valuation { estate: this_valuation_estate } AS this_valuation
    RETURN head(collect(this_valuation)) AS this_valuation
}
RETURN this { valuation: this_valuation } AS this

SKIP $param7
LIMIT $param8
Params:
{
  "param0": 0,
  "param1": 0,
  "param2": {
    "low": 0,
    "high": 0
  },
  "param3": [
    "APARTMENT"
  ],
  "param4": [
    "13001"
  ],
  "param5": {
    "low": 20,
    "high": 0
  },
  "param6": {
    "low": 40,
    "high": 0
  },
  "param7": {
    "low": 20,
    "high": 0
  },
  "param8": {
    "low": 40,
    "high": 0
  }
}
```

**Expected behavior**
Out of 26 total results, I would expect 6 results with `"offset": 20` and `"limit": 40`.

**System (please complete the following information):**

* OS: Manjaro Linux 22.0.0 Sikaris
* Version: 3.12.0
* Node.js version: v16.16.0

**Comment by CraigDeverall** [comment]
1+ this issue :)

**Comment by Tenrys** [comment]
Any news? This issue seems pretty critical for pagination, the first guess is that the `SKIP` and `LIMIT` directives are applied as many times as there are relationships queried in the query.

**Comment by Tanner-Scadden** [comment]
I am having issues with pagination as well after the first page of results.

**Comment by jsegura** [comment]
1+ this issue :)

We are also facing the duplicated `SKIP x LIMIT x` output in cypher query. And only happens when adding relationship fields to the query.

**Comment by tonxxd** [comment]
+1

**Comment by darrellwarde** [comment]
All - apologies about this. We use a GitHub Project to manage bug triage, and because the "bug report" label was added retrospectively, this issue didn't get added to the Project and seems to have unfortunately been overlooked. I'll take a look now.

**Comment by neo4j-team-graphql** [comment]
This bug report has been assigned high priority to fix. If you wish to contribute a fix, please branch from `master` and submit your PR with the base set to `master`. Thanks!

**Comment by darrellwarde** [comment]
This should now be fixed in 3.13.0 - please let us know if you're still having issues!
